litestar/typing.py

```diff
@@ -397,7 +397,7 @@ def generic_types(self) -> tuple[FieldDefinition, ...] | None:
     def is_dataclass_type(self) -> bool:
         """Whether the annotation is a dataclass type or not."""
 
-        return is_dataclass(self.origin or self.annotation)
+        return is_dataclass(cast("type", self.origin or self.annotation))
```

Comment:
The cast is necessary b/c `self.origin or self.annotation` is typed `Any` b/c `self.origin` and `self.annotation` are both typed `Any`.

I think we should be able to change the type of `origin` to `type[Any] | None` because where `annotation` is one of the type alias things, it will always return a class.

Then, I'm thinking I will add a method `FieldDefinition.origin_or_annotation()` which will be typed to return `type[Any]`. We use the `field_def.origin or field_def.annotation` pattern quite a bit, and whenever `annotation` is not a type (ie. special form thing, or whatever) then origin will be non-none, otherwise, `annotation` will always be a type.

I'll do that separate though.

Reply:
Oh so this is why mypy was complaining.

Reply:
Yeah - the overloads for `is_dataclass` don't have a variant for `Any` arg: https://github.com/python/typeshed/blob/ba3f23b2040927cb278dcb4d5ea5f8c43eb5c8b3/stdlib/dataclasses.pyi#L226-L230